File: python/decision_engine.py
"""Decision Engine - combines all analyses into a final Buy/Hold/Sell recommendation."""


class DecisionEngine:
    """Weighted scoring system that merges TA, FA, and News signals."""

    WEIGHTS = {
        "technical": 0.40,
        "fundamental": 0.35,
        "news": 0.25,
    }

    def __init__(self, ta_result: dict, fa_result: dict, news_result: dict):
        self.ta = ta_result
        self.fa = fa_result
        self.news = news_result

    def decide(self) -> dict:
        ta_score = self._category_score(self.ta.get("signals", {}))
        fa_score = self._category_score(self.fa.get("signals", {}))
        news_score = self._category_score(self.news.get("signals", {}))

        weighted = (
            ta_score * self.WEIGHTS["technical"]
            + fa_score * self.WEIGHTS["fundamental"]
            + news_score * self.WEIGHTS["news"]
        )

        recommendation, confidence = self._map_score(weighted)

        return {
            "scores": {
                "Technical Score": round(ta_score, 2),
                "Fundamental Score": round(fa_score, 2),
                "News Score": round(news_score, 2),
                "Weighted Score": round(weighted, 2),
            },
            "weights": self.WEIGHTS,
            "recommendation": recommendation,
            "confidence": confidence,
            "explanation": self._explain(ta_score, fa_score, news_score, weighted, recommendation),
        }

    @staticmethod
    def _category_score(signals: dict) -> float:
        """Convert signals dict to a -1..+1 score."""
        if not signals:
            return 0.0
        total = 0
        count = 0
        for key, (direction, _) in signals.items():
            if key.startswith("_") and key.endswith("Overall"):
                continue  # skip aggregate keys to avoid double-counting
            count += 1
            if direction == "Bullish":
                total += 1
            elif direction == "Bearish":
                total -= 1
        return total / count if count else 0.0

    @staticmethod
    def _map_score(score: float) -> tuple[str, str]:
        if score >= 0.4:
            return "STRONG BUY", "High"
        if score >= 0.15:
            return "BUY", "Moderate"
        if score > -0.15:
            return "HOLD", "—"
        if score > -0.4:
            return "SELL", "Moderate"
        return "STRONG SELL", "High"

    @staticmethod
    def _explain(ta: float, fa: float, news: float, weighted: float, rec: str) -> str:
        parts = []
        parts.append(f"Technical analysis score: {ta:+.2f} (weight 40%)")
        parts.append(f"Fundamental analysis score: {fa:+.2f} (weight 35%)")
        parts.append(f"News sentiment score: {news:+.2f} (weight 25%)")
        parts.append(f"Combined weighted score: {weighted:+.2f}")
        parts.append(f"Recommendation: {rec}")
        parts.append("")
        parts.append("Score interpretation: -1.0 (Strong Sell) to +1.0 (Strong Buy)")
        parts.append("  >= +0.40  STRONG BUY  | >= +0.15  BUY  | -0.15 to +0.15  HOLD")
        parts.append("  <= -0.15  SELL        | <= -0.40  STRONG SELL")
        return "\n".join(parts)

File: csharp/StockAnalysis/DecisionEngine.cs
namespace StockAnalysis;

/// <summary>
/// Weighted scoring system that merges TA, FA, and News signals into a final recommendation.
/// </summary>
public class DecisionEngine
{
    private static readonly Dictionary<string, double> Weights = new()
    {
        ["Technical"] = 0.40,
        ["Fundamental"] = 0.35,
        ["News"] = 0.25,
    };

    private readonly AnalysisResult _ta;
    private readonly AnalysisResult _fa;
    private readonly NewsResult _news;

    public DecisionEngine(AnalysisResult ta, AnalysisResult fa, NewsResult news)
    {
        _ta = ta;
        _fa = fa;
        _news = news;
    }

    public DecisionResult Decide()
    {
        double taScore = CategoryScore(_ta.Signals);
        double faScore = CategoryScore(_fa.Signals);
        double newsScore = CategoryScore(_news.Signals);

        double weighted = taScore * Weights["Technical"]
                        + faScore * Weights["Fundamental"]
                        + newsScore * Weights["News"];

        var (rec, conf) = MapScore(weighted);

        return new DecisionResult
        {
            Scores = new Dictionary<string, double>
            {
                ["Technical Score"] = Math.Round(taScore, 2),
                ["Fundamental Score"] = Math.Round(faScore, 2),
                ["News Score"] = Math.Round(newsScore, 2),
                ["Weighted Score"] = Math.Round(weighted, 2),
            },
            Weights = Weights,
            Recommendation = rec,
            Confidence = conf,
            Explanation = Explain(taScore, faScore, newsScore, weighted, rec),
        };
    }

    private static double CategoryScore(Dictionary<string, Signal> signals)
    {
        if (signals.Count == 0) return 0.0;
        int total = 0, count = 0;
        foreach (var (key, signal) in signals)
        {
            if (key.StartsWith("_") && key.EndsWith("Overall")) continue;
            count++;
            if (signal.Direction == "Bullish") total++;
            else if (signal.Direction == "Bearish") total--;
        }
        return count > 0 ? (double)total / count : 0.0;
    }

    private static (string Recommendation, string Confidence) MapScore(double score) => score switch
    {
        >= 0.4 => ("STRONG BUY", "High"),
        >= 0.15 => ("BUY", "Moderate"),
        > -0.15 => ("HOLD", "—"),
        > -0.4 => ("SELL", "Moderate"),
        _ => ("STRONG SELL", "High"),
    };

    private static string Explain(double ta, double fa, double news, double weighted, string rec)
    {
        var lines = new[]
        {
            $"Technical analysis score: {ta:+0.00;-0.00} (weight 40%)",
            $"Fundamental analysis score: {fa:+0.00;-0.00} (weight 35%)",
            $"News sentiment score: {news:+0.00;-0.00} (weight 25%)",
            $"Combined weighted score: {weighted:+0.00;-0.00}",
            $"Recommendation: {rec}",
            "",
            "Score interpretation: -1.0 (Strong Sell) to +1.0 (Strong Buy)",
            "  >= +0.40  STRONG BUY  | >= +0.15  BUY  | -0.15 to +0.15  HOLD",
            "  <= -0.15  SELL        | <= -0.40  STRONG SELL",
        };
        return string.Join(Environment.NewLine, lines);
    }
}
